[VPlan] Use VPlan operand order for VPBlendRecipes.

fhahn opened the pull request:
Remove use of IR references and the order of incoming values of IR phis when creating VPBlendRecipes. Instead, simply use the incoming operands and blocks from the VPWidenPHIRecipe.

Note that this changes the order of the incoming operands/masks for some blends.

fhahn committed:
[VPlan] Use VPlan operand order for VPBlendRecipes.

Remove use of IR references and the order of incoming values of IR phis
when creating VPBlendRecipes. Instead, simply use the incoming operands
and blocks from the VPWidenPHIRecipe.

Note that this changes the order of the incoming operands/masks for some
blends.

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8538,36 +8538,30 @@ VPWidenIntOrFpInductionRecipe *VPRecipeBuilder::tryToOptimizeInductionTruncate(
-  // Map incoming IR BasicBlocks to incoming VPValues, for lookup below.
-  // TODO: Add operands and masks in order from the VPlan predecessors.
-  DenseMap<BasicBlock *, VPValue *> VPIncomingValues;
-  for (const auto &[Idx, Pred] : enumerate(predecessors(Phi->getParent())))
```
The operands of VPWidenPHIRecipe are set according to the order of the original predecessors of the underlying Phi, so unclear why this change affects test?

fhahn replied:
They were set according to the order of incoming values/blocks in the original phi, which in some cases can be different to the order of predecessors in LLVM IR, which is causing the test changes.

fhahn replied:
Thanks, I'll land the NFC refactorings first, threading teh recipes here, but still preserving the original order, as this keeps the overall diff smaller, thanks!

ayalz commented on llvm/test/Transforms/LoopVectorize/AArch64/blend-costs.ll:
```diff
@@ -219,7 +219,7 @@ define void @test_blend_feeding_replicated_store_2(ptr noalias %src, ptr %dst, i
 ; CHECK-NEXT:    [[TMP6:%.*]] = select <16 x i1> [[TMP4]], <16 x i1> [[TMP5]], <16 x i1> zeroinitializer
 ; CHECK-NEXT:    [[TMP7:%.*]] = or <16 x i1> [[TMP6]], [[TMP3]]
-; CHECK-NEXT:    [[PREDPHI:%.*]] = select <16 x i1> [[TMP6]], <16 x i8> zeroinitializer, <16 x i8> splat (i8 1)
+; CHECK-NEXT:    [[PREDPHI:%.*]] = select <16 x i1> [[TMP3]], <16 x i8> splat (i8 1), <16 x i8> zeroinitializer
```
Select operands exchange order and condition changes from TMP6=(!TMP3)?TMP5:0 to TMP3?

fhahn replied:
yep

ayalz replied:
So entries of TMP3 that are true continue to set PREDPHI to i8 1, but entries of TMP3 that are false originally set PREDPHI to either 0 or 1 according to TMP5 (which is !%c.0 if followed correctly), but now sets it to 0 regardless of TMP5?

fhahn replied:
Yep, but we only use i8 1 when TMP3 is true and i8 0 zero otherwise. When TMP3 is false, PREDPHI is not used when %c.0 is true.

fhahn committed:
Merge remote-tracking branch 'origin/main' into blend-from-vpwidenphi

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8538,36 +8538,30 @@ VPWidenIntOrFpInductionRecipe *VPRecipeBuilder::tryToOptimizeInductionTruncate(
-    OperandsWithMask.push_back(VPIncomingValues.lookup(Pred));
-    VPValue *EdgeMask = getEdgeMask(Pred, Phi->getParent());
+    const VPBasicBlock *Pred = PhiR->getIncomingBlock(In);
+    OperandsWithMask.push_back(PhiR->getIncomingValue(In));
```
reordered, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8538,36 +8538,30 @@ VPWidenIntOrFpInductionRecipe *VPRecipeBuilder::tryToOptimizeInductionTruncate(
-    VPValue *EdgeMask = getEdgeMask(Pred, Phi->getParent());
+    const VPBasicBlock *Pred = PhiR->getIncomingBlock(In);
+    OperandsWithMask.push_back(PhiR->getIncomingValue(In));
+    VPValue *EdgeMask = getEdgeMask(Pred, PhiR->getParent());
```
Yep, in preparation for VPPredicator, I sunk accesses of `VPBB2IRBB` to `VPRecieBuilder` in cfde685e22d7babe0693fd4daf90e3bc20e98ec0

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8954,15 +8948,18 @@ bool VPRecipeBuilder::getScaledReductions(
+  Instruction *Instr = R->getUnderlyingInstr();
+  SmallVector<VPValue *, 4> Operands(R->operands());
+  if (auto *PhiR = dyn_cast<VPWidenPHIRecipe>(R)) {
+    if (PhiR->getParent()->getNumPredecessors() != 0)
```
Yep, added a comment that this is to handle phis in non-header blocks, thanks

ayalz replied:
This works in HCFG mode, excluding VPlan's entry block - which should be free of VPWidenPHIRecipes. Clearer to ask explicitly if PhiR's parental block differs from the entry of it enclosing parental region?

fhahn replied:
Updated, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h:
```diff
@@ -125,7 +125,7 @@ class VPRecipeBuilder {
   /// Handle non-loop phi nodes. Return a new VPBlendRecipe otherwise. Currently
```
Updated, thanks

fhahn committed:
!fixup address latest comments, thanks

ayalz commented on llvm/test/Transforms/LoopVectorize/AArch64/blend-costs.ll:
```diff
@@ -28,7 +28,7 @@ define void @test_blend_feeding_replicated_store_1(i64 %N, ptr noalias %src, ptr
 ; CHECK-NEXT:    [[TMP8:%.*]] = xor <16 x i1> [[TMP6]], splat (i1 true)
 ; CHECK-NEXT:    [[TMP9:%.*]] = or <16 x i1> [[TMP7]], [[TMP8]]
-; CHECK-NEXT:    [[PREDPHI:%.*]] = select <16 x i1> [[TMP7]], <16 x ptr> [[BROADCAST_SPLAT]], <16 x ptr> zeroinitializer
+; CHECK-NEXT:    [[PREDPHI:%.*]] = select <16 x i1> [[TMP6]], <16 x ptr> [[BROADCAST_SPLAT]], <16 x ptr> zeroinitializer
```
Is this change ok - from TMP7 (=TMP6?0:0, i.e., =0) to TMP6 (which may include non-zeroes)?

fhahn replied:
Yes, this is due to the second condition being constant `false`. We are only using the broadcast if the original condition (TMP6) is true, otherwise we don't reach the user due to constant `false`

fhahn committed:
Merge remote-tracking branch 'origin/main' into blend-from-vpwidenphi

fhahn committed:
!fixup check enclosing loop region entry.

ayalz reviewed (approved):
This LGTM, although it might be better to isolate the operand reordering part into a separate (non NFC) patch dedicated to it alone.

fhahn committed:
Merge remote-tracking branch 'origin/main' into blend-from-vpwidenphi

fhahn committed:
[VPlan] Construct VPBlendRecipe from VPWidenPHIRecipe (NFC).

Update VPRecipeBuilder to construct VPBlendRecipe from VPWidenPHIRecipe,
starting to thread recipes through the builder instead of the
underlying IR instruction up-front.

Landing first part of approved
https://github.com/llvm/llvm-project/pull/139475 separately as NFC as
suggested.

fhahn committed:
Merge branch 'blend-from-vpwidenphi-keep-order' into blend-from-vpwidenphi

fhahn committed:
Merge remote-tracking branch 'origin/main' into blend-from-vpwidenphi
